src/thermo/ThermoPhase.cpp

```diff
@@ -252,11 +255,28 @@ void ThermoPhase::setState(const AnyMap& input_state)
         }
         state.erase("Y");
     }
-    // set thermodynamic state using whichever property pair is found
+    // set thermodynamic state using whichever property set is found
     if (state.size() == 0) {
         setState_TP(298.15, OneAtm);
     } else if (state.hasKey("T") && state.hasKey("P")) {
-        setState_TP(state.convert("T", "K"), state.convert("P", "Pa"));
+        double T = state.convert("T", "K");
+        double P = state.convert("P", "Pa");
+        if (state.hasKey("Q")) {
+            double Q = state["Q"].asDouble();
+            double Psat = satPressure(T);
+            if (std::abs(Psat / P - 1) < 1e-6) {
```

Comment:
The cython version 
https://github.com/Cantera/cantera/blob/358195ea9eb99ccbf356338b7d31fac55fc0eb01/interfaces/cython/cantera/thermo.pyx#L1762-L1792
uses slightly different tolerances here (via [np.isclose](https://docs.scipy.org/doc/numpy/reference/generated/numpy.isclose.html)). Would it make sense to move all of this to a single spot in the C++ layer? (i.e. a new `setState_TPQ`; that way, the unit tests would apply universally). 

**PS:** I do realize that `Q` is still `X` in the C++ layer, which should be addressed eventually (together with `D` vs `R`).

Reply:
Yeah, that's a good point on just having a single implementation of this procedure.

I'm not really convinced on the cost/benefit of deprecating and renaming all the relevant C++ functions.

Reply:
limited cost / limited benefit = not a high priority. 

However, in this case, `setState_TPX` is already taken and we'd be starting to mix nomenclatures (i.e. `Q` and `D` are starting to bleed more and more into the C++ layer).

Reply:
One fun question that hasn't been resolved in either implementation: what should happen if T is above the critical point? Right now, it throws an exception while trying to evaluate the saturation pressure.

Reply:
Ha! `Q` would be immaterial, but it's definitely a bug!

Reply:
I would certainly agree for any value 0 < Q < 1, but if a user were to specify water at T = 700 K, 1000 Pa and Q = 1, it doesn't seem that different from any other superheated state, which we do allow.

Reply:
I guess it depends on whether you describe a supercritical fluid as liquid or vapor ... from a strict perspective, it is neither.

Reply:
At least some definitions of a supercritical fluid are that both temperature and pressure need to be above the critical point. In this case, P > Pc with T < Tc is just a liquid and should be allowed with Q = 0, and correspondingly, P < Pc with T > Tc is just a gas / vapor and should be allowed with Q = 1.

Reply:
I agree on what you said above. Above the vapor dome Q is somewhat murky (in particular, above the critical point at equal v, properties of a 'liquid' and a 'vapor' are the same; I checked that vc is *suggested* as a demarcation although I'd rather not use different labels). From an implementation viewpoint, your suggestion makes sense. :+1:

Reply:
Yes, I agree that it's murky in the supercritical region, but I think we should accept some value for Q there so that you can, for example, round-trip `SolutionArray` using T-P-Q as the columns even in the supercritical region. I opted for just ignoring the value of Q when P > Pc and T > Tc, rather than imposing a particular dividing line. This is implemented in the latest commit.